mountpoint-s3/src/prefetch/parquet_stream.rs

```diff
@@ -0,0 +1,423 @@
+        }
+    }
+}
+
+/// Get from client as requested
+async fn get_from_client<Client>(
+    client: &Client,
+    bucket: &str,
+    id: &ObjectId,
+    range: RequestRange,
+) -> Result<Vec<Part>, PrefetchReadError<Client::ClientError>>
+where
+    Client: ObjectClient + Send + Sync + 'static,
+{
+    let key = id.key();
+    let block_aligned_byte_range = range;
+
+    trace!(?key, range =? block_aligned_byte_range, "fetching data from client");
+
+    let get_object_result = client
+        .get_object(
+            bucket,
+            key,
+            Some(block_aligned_byte_range.into()),
+            Some(id.etag().clone()),
+        )
+        .await
+        .map_err(PrefetchReadError::GetRequestFailed)?;
+
+    pin_mut!(get_object_result);
+
+    let mut parts = Vec::new();
+    let mut offset = range.start();
+
+    while let Some(result) = get_object_result.next().await {
+        match result {
+            Ok((part_offset, body)) => {
+                trace!(offset, length = body.len(), "received GetObject part");
+                metrics::counter!("s3.client.total_bytes", "type" => "read").increment(body.len() as u64);
+
+                if part_offset != offset {
+                    warn!(key, part_offset, offset, "wrong offset for GetObject body part");
+                    return Err(PrefetchReadError::GetRequestReturnedWrongOffset {
+                        offset: part_offset,
+                        expected_offset: offset,
+                    });
+                }
+
+                let body_len = body.len();
+                let part = Part::new(id.clone(), offset, ChecksummedBytes::new(body.into()));
+                parts.push(part);
+                offset += body_len as u64;
+            }
+            Err(e) => {
+                warn!(key, error=?e, "GetObject body part failed");
+                return Err(PrefetchReadError::GetRequestFailed(e));
+            }
+        }
+    }
+
+    trace!("request finished");
+    Ok(parts)
+}
+
+/// Get row groups and columns that intersect with the given range
+///
+/// Returns a vector of tuples containing the row group and column index, and the intersecting range
+fn get_row_groups_and_columns(
+    interval_tree: &IntervalTree<u64, (RowGroupIndex, ColumnIndex)>,
+    range: Range<u64>,
+) -> Vec<((RowGroupIndex, ColumnIndex), Range<u64>)> {
+    interval_tree
+        .iter_sorted()
+        .filter_map(|element| {
+            let intersection_start = element.range.start.max(range.start);
+            let intersection_end = element.range.end.min(range.end);
+
+            if intersection_start < intersection_end {
+                Some((element.value, intersection_start..intersection_end))
+            } else {
+                None
+            }
+        })
+        .take_while(|(_, intersect_range)| intersect_range.start < range.end)
+        .collect()
+}
+
+/// Merge a new range with the existing cache
+fn merge_ranges(
```

Comment:
This method is doing a lot of data manipulation. Is it really needed? Could we get by with some data duplication and/or a less compact column cache?